src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.cs

```diff
@@ -30,6 +30,26 @@ public sealed partial class SqlNotificationRequest
 }
 namespace Microsoft.Data.SqlClient
 {
+    /// <include file='../../../../doc/snippets/Microsoft.Data.SqlClient/ActiveDirectoryAuthenticationProvider.xml' path='docs/members[@name="ActiveDirectoryAuthenticationProvider"]/ActiveDirectoryAuthenticationProvider/*'/>
+    public sealed partial class ActiveDirectoryAuthenticationProvider : SqlAuthenticationProvider
+    {
+        /// <include file='../../../../doc/snippets/Microsoft.Data.SqlClient/ActiveDirectoryAuthenticationProvider.xml' path='docs/members[@name="ActiveDirectoryAuthenticationProvider"]/ctor/*'/>
+        public ActiveDirectoryAuthenticationProvider() { }
+        /// <include file='../../../../doc/snippets/Microsoft.Data.SqlClient/ActiveDirectoryAuthenticationProvider.xml' path='docs/members[@name="ActiveDirectoryAuthenticationProvider"]/ctor2/*'/>
+        public ActiveDirectoryAuthenticationProvider(System.Func<Microsoft.Identity.Client.DeviceCodeResult, System.Threading.Tasks.Task> deviceCodeFlowCallbackMethod) { }
+        /// <include file='../../../../doc/snippets/Microsoft.Data.SqlClient/ActiveDirectoryAuthenticationProvider.xml' path='docs/members[@name="ActiveDirectoryAuthenticationProvider"]/AcquireTokenAsync/*'/>
+        public override System.Threading.Tasks.Task<SqlAuthenticationToken> AcquireTokenAsync(SqlAuthenticationParameters parameters) { throw null; }
+        /// <include file='../../../../doc/snippets/Microsoft.Data.SqlClient/ActiveDirectoryAuthenticationProvider.xml' path='docs/members[@name="ActiveDirectoryAuthenticationProvider"]/SetDeviceCodeFlowCallback/*'/>
+        public void SetDeviceCodeFlowCallback(System.Func<Microsoft.Identity.Client.DeviceCodeResult, System.Threading.Tasks.Task> deviceCodeFlowCallbackMethod) { }
+        /// <include file='../../../../doc/snippets/Microsoft.Data.SqlClient/ActiveDirectoryAuthenticationProvider.xml' path='docs/members[@name="ActiveDirectoryAuthenticationProvider"]/SetCustomWebUI/*'/>
+        public void SetCustomWebUI(Microsoft.Identity.Client.Extensibility.ICustomWebUi customWebUi) { }
```

Comment:
Having Microsoft.Identity.Client.Extensibility.ICustomWebUi as part of our public API feels like the future may hold the same problem that we have in the AKV provider API, which we are now breaking. However, I'm not sure creating our own interface will save us from much future pain if MSAL decides to change theirs. Ideas? Thought?

Reply:
We can do that by accepting delegate/callback for [AcquireAuthorizationCodeAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.extensibility.icustomwebui.acquireauthorizationcodeasync?view=azure-dotnet) and pass it internally with CustomWebUi instance?

Reply:
It would save us from API changes, but if their whole design changes some day, we'll have to drop this API based on MSAL version upgrade, but so far it doesn't seem to go that far.

Reply:
I updated the PR accordingly, it does open more possibilities for a more generic implementation.